proof/refine/ARM/Schedule_R.thy

```diff
@@ -4210,6 +4205,920 @@ lemma refillUnblockCheck_corres':
+  apply (simp add: getCurSc_def)
+  apply (wpsimp wp: hoare_return_sp)
+  done
+
+lemma refillBudgetCheckRoundRobin_corres:
+  "corres dc
+          (cur_sc_active and (\<lambda>s. sc_at (cur_sc s) s))
+          ((\<lambda>s'. valid_refills' (ksCurSc s') s') and (\<lambda>s'. sc_at' (ksCurSc s') s'))
+          (refill_budget_check_round_robin usage) (refillBudgetCheckRoundRobin usage)"
+  supply projection_rewrites[simp]
+  apply (subst is_active_sc_rewrite)
+  apply (clarsimp simp: refill_budget_check_round_robin_def refillBudgetCheckRoundRobin_def)
+  apply (rule corres_split'[rotated 2, OF gets_sp getCurSc_sp])
+   apply (corressimp corres: getCurSc_corres)
+  apply (rule_tac Q="\<lambda>s. is_active_sc' (ksCurSc s) s" in corres_cross_add_guard)
+   apply (rule_tac ptr="ksCurSc s'" in is_active_sc'_cross[OF state_relation_pspace_relation]; simp)
+   apply clarsimp
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF updateRefillHd_corres], simp)
+       apply (clarsimp simp: refill_map_def)
+      apply (rule updateRefillTl_corres, simp)
+      apply (clarsimp simp: refill_map_def)
+     apply (wpsimp simp: update_refill_hd_rewrite wp: set_refills_wp get_refills_wp)
+    apply (wpsimp wp: hoare_vcg_conj_lift)
+     apply (wpsimp simp: updateRefillHd_def wp: updateSchedContext_wp)
+    apply (wpsimp wp: updateRefillHd_valid_objs')
+   apply (clarsimp simp: obj_at_def is_active_sc2_def is_sc_obj opt_map_red
+                  split: option.split_asm Structures_A.kernel_object.split_asm)
+  apply (clarsimp simp: obj_at_simps fun_upd_def[symmetric] scBits_simps ps_clear_upd)
+  apply (clarsimp simp: is_active_sc'_def valid_obj'_def valid_sched_context'_def valid_refills'_def
+                 split: option.split_asm)
+  done
+
+lemma head_insufficient_length_greater_than_one:
+  "\<lbrakk>the (head_insufficient sc_ptr s);
+    pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr\<rbrakk>
+   \<Longrightarrow> pred_map (\<lambda>cfg. Suc 0 < length (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr"
+  apply (frule head_insufficient_true_imp_insufficient)
+   apply (clarsimp simp: vs_all_heap_simps)
+  apply (clarsimp simp: vs_all_heap_simps sc_at_ppred_def refills_unat_sum_def word_less_nat_alt)
+  apply (case_tac "sc_refills y"; fastforce dest!: member_le_sum_list)
+  done
+
+lemma length_sc_refills_cross:
+  "\<lbrakk>(s, s') \<in> state_relation; sc_at scp s; sc_at' scp s'; valid_refills' scp s'\<rbrakk>
+   \<Longrightarrow> ((\<lambda>sc. P (length (sc_refills sc))) |< scs_of2 s) scp
+        = ((\<lambda>sc'. P (scRefillCount sc')) |< scs_of' s') scp"
+  apply (drule (2) state_relation_sc_relation)
+  apply (clarsimp simp: obj_at_simps is_sc_obj valid_refills'_def sc_relation_def opt_map_red)
+  done
+
+lemma update_refill_hd_rewrite:
+  "update_refill_hd scPtr (f \<circ> g)
+   = do update_refill_hd scPtr g;
+        update_refill_hd scPtr f
+     od"
```

Comment:
Is this a duplicate? A lemma by the same name seems to be used earlier in the file, so I think it's either an accidental duplication or a name clash.

Reply:
Yep, name clash. I've renamed it to `update_refill_hd_comp`